---
title: "Integrative annotation for human genetics: Approaches with Bioconductor"
author: "Vincent J. Carey"
date: "`r doc_date()`"
abstract: >
  The study of genetic variation among human individuals
  engenders substantial archives of details on individual genomes.
  Details of genetic variation are interpreted using evolving
  resources assembled at various scales.  NCBI's dbSNP assembles
  information on relatively small variants that occur with
  some regularity in human populations.  The EMBL-EBI GWAS catalog
  provides a curated view of replicated studies of genotype-phenotype
  association.  This workflow reviews the use of Bioconductor packages
  that help investigators interpret their own findings in studies
  of human genetics.
vignette: >
  %\VignetteIndexEntry{Integrative annotation for human genetics: Approaches with Bioconductor}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  BiocStyle::html_document
---

```{r setup,echo=FALSE,results="hide"}
suppressPackageStartupMessages({
library(gwascat)
library(AnnotationHub)
library(DT)
})
```

# Background

The table of contents of Vogel and Motulsky's [_Human Genetics: Problems and Approaches_](https://books.google.com/books?id=xuztCAAAQBAJ&lpg=PA6&dq=human%20genetics&pg=PR32#v=onepage&q=human%20genetics&f=false)
is a worthy survey of concepts addressed in research on human
genetics and genetic medicine.  The frontiers of knowledge in the field
are shifting, and expectations are high.

In this workflow, I aim to show how researchers can use R
to interrogate important resources of use in human genetic
epidemiology and medical genomics.  I show how to program with two
genome-wide association study
(GWAS) catalogs, 
the [EMBL-EBI GWAS catalog](https://www.ebi.ac.uk/gwas/) 
and the [NHLBI GRASP v2.0](http://iapps.nhlbi.nih.gov/GRASP/Overview.aspx).
Aspects of findings reported in these studies are then
integrated with new functional and structural annotation resources
to aid in variant interpretation.
An appendix provides
brief treatment of "reference genome builds" for _Homo sapiens_,
packages for querying contents of
the [NCBI dbSNP](http://www.ncbi.nlm.nih.gov/SNP/),
and tools for obtaining and programming with gene models.

# The gwascat package for the EMBL-EBI GWAS catalog

## Basic operations, fields, and interactive tabulation 

The EMBL-EBI catalog is presented using GRCh38 coordinates.
```{r lkg}
library(gwascat)
data(ebicat38)
ebicat38
```

While there are `r length(ebicat38)` records, the number of unique loci is
```{r lku}
length(unique(ebicat38$SNPS))
```

A full view of the metadata about each study result is available
with the commands
```{r eval=FALSE}
library(DT)
datatable(as.data.frame(mcols(ebicat38)), options=list(autoWidth=TRUE,
  style="height:30px"), pageLength=5)
```

The following command generates
a table restricting attention to records related to asthma.

```{r gett}
suppressWarnings({
aind = grep("sthma", ebicat38$DISEASE.TRAIT)
datatable(as.data.frame(mcols(ebicat38[aind])), options=list(autoWidth=TRUE,
  style="height:30px", pageLength=5))
})
```

## Navigating traits using the EMBL-EBI Experimental Factor Ontology

Field `MAPPED_TRAIT_URI` includes a comma-delimited string with
URIs referring to an ontology for traits and other factors relevant
to biological experiments and observations.  The underlying
ontology is available in the form of an annotated algebraic graph.

```{r lkggg}
data(efo.obo.g)
efo.obo.g
```
There are over 16000 terms in the ontology.  Terms and term-related
metadata are manipulated using methods of the `r Biocpkg("graph")`
package.
```{r lkgt}
nodes(efo.obo.g)[1:4] # imported directly from OBO
names(nodeData(efo.obo.g)[[1]])
sapply(nodeData(efo.obo.g)[1:4], "[[", "name")
```

Let's obtain the EFO annotation for SNP `rs347412`.
```{r lkefo}
ind = which(ebicat38$SNPS == "rs347412")
urs = ebicat38$MAPPED_TRAIT_URI[ind]
urs
```
These entries must be converted to match the EFO OBO node
nomenclature.  We then find the EFO names of the factors annotated
to this SNP.
```{r lkefo2}
nn = uri2node(urs)
nd = nodeData(efo.obo.g, nn)
sapply(nd, "[[", "name")
```

The current representation of the ontology is a directed graph
with links pointing from a term to its semantic parent.  We
convert to an undirected graph to explore semantic neighborhoods of terms.
The `adj` method will return the nodes adjacent to a specified node.
Here we obtain the terms accessible from `respiratory system disease`
with a single step.
```{r explsem}
rsdn = adj(ugraph(efo.obo.g), "EFO:0000684")  # respiratory system disease
unlist(sapply(nodeData(efo.obo.g, rsdn[[1]]), "[[", "name"))
```
The `r Biocpkg("RBGL")` package can be used to deploy diverse graph algorithms
against this ontology.

Once a node name of interest has been found, `node2uri` can be used
with code to find
GWAS hits deemed relevant by the curators.

```{r lkuri}
e270 = ebicat37[ grep(node2uri("EFO:0000270"), ebicat37$MAPPED_TRAIT_URI) ]
length(e270)
table(e270$DISEASE.TRAIT)[1:5]
```

# Appendix: Bioconductor infrastructure supporting genetic data analysis 

## Reference builds of the human genome sequence

The most recent build of the human genomic sequence
is labeled GRCh38.  Using Bioconductor, we can be very concrete about what this
is.

```{r lkhg}
library(BSgenome.Hsapiens.NCBI.GRCh38)
class(Hsapiens)
Hsapiens
class(Hsapiens$"1")
Hsapiens$"1"
```
